### DISCUSSION TITLE
GraphQL API v4 - Accessing RepositoryVulnerabilityAlert

### DISCUSSION THREAD

**Discussion by sbhzi** [community/community#24413] [Apps, API and Webhooks]
<p>So I’m trying to access the GraphQL API v4 as such:</p>
<pre><code>import os
import requests
import json

url = 'https://api.github.com/graphql'
query = """
  query someFunction($repository_name: String!){
    repositoryVulnerabilityAlert {
      repository(name: $repository_name) {
        name
      }

    }
  }
"""
variables = """
{
  "repository_name":"my_repo"
}
"""
api_token = os.environ['bearer_token']
payload = {'Accept': 'application/vnd.github.vixen-preview+json'}
headers = {'Authorization': 'bearer %s' % api_token, 'Accept': 'application/vnd.github.vixen-preview+json'}

r = requests.post(url=url, json={'query': query, 'variables': variables}, data=json.dumps(payload), headers=headers)

print (r.text)
</code></pre>
<p>And I keep receiving:</p>
<pre><code>{"errors":[{"message":"A query attribute must be specified and must be a string."}]}
</code></pre>
<p>What am I doing wrong?</p>

**Comment by nickvanw** [comment, marked as answer]
<p>Hey <a class="mention" href="/sgript">@sgript</a>!</p>
<p>I see a few mistakes in your script, both in the GraphQL query you’re running, as well as in how you’re running it. </p>
<p>The GraphQL Query that you want in order to list repository vulnerability alerts off of a single repo should look something like this:</p>
<pre><code>query repoVulns($owner:String!, $name:String!){
  repository(owner:$owner, name:$name) {
    vulnerabilityAlerts(first:10) {
      nodes {
        id
      }
    }
  }
}
</code></pre>
<p>In addition, it looks like you’re passing the wrong data in with <code>data=json.dumps(payload)</code>, which seems to be a variable that only has some headers in it. I went ahead and made the changes, and I think this shoud work:</p>
<pre><code>import os
import requests
import json

url = 'https://api.github.com/graphql'
query = """
query repoVulns($owner:String!, $name:String!){
  repository(owner:$owner, name:$name) {
    vulnerabilityAlerts(first:10) {
      nodes {
        id
      }
    }
  }
}

"""
variables = """
{"owner": "owner", "name": "name"}
"""
api_token = os.environ['bearer_token']
headers = {'Authorization': 'Bearer %s' % api_token, 'Accept': 'application/vnd.github.vixen-preview+json'}

r = requests.post(url=url, json={'query': query, 'variables': variables}, headers=headers)

print (r.text)
</code></pre>
<p>You’ll have to change the owner and the name to the Repository owner and name that you’re trying to look up, but I think this should work!</p>
<p>Let me know!</p>

**Comment by sbhzi** [comment]
<p>Hi <a class="mention" href="/nickvanw">@nickvanw</a> !</p>
<p>Thanks for your prompt response, I appreciate it.</p>
<p>I’ve used the code suggested and added in the variables assigned to <code>owner</code> and <code>name</code> for the repository. Although I’m receiving an actual data response now, I receive an empty node, as such:</p>
<p><code>{"data":{"repository":{"vulnerabilityAlerts":{"nodes":[]}}}}</code></p>
<p>It’s worth stating that the repository I’m trying to target is apart of a team and hence I’ve passed the team’s name as the value assigned to the <code>owner</code> variable, being that the url of our repo reads <a href="http://github.com/theTeam/theRepo" class="inline-onebox-loading" rel="nofollow noopener">github.com/theTeam/theRepo</a></p>
<p>And I’ve simply assigned <code>theTeam</code> to <code>owner</code> and <code>theRepo</code> to <code>name</code>.</p>
<p>We’ve checked and we definitely have vulnerability alerts listed. Any ideas as to why it may be returning blank would be very welcome!</p>

**Comment by sbhzi** [comment]
<p>An update on this <a class="mention" href="/nickvanw">@nickvanw</a> </p>
<p>It seems to have been a permissions issue - in order to run this particular call, I needed to be listed as the owner of the repository. After having a colleague try to run the script (who happened to be the owner of the repository), with their bearer token, some results were returned.</p>
<p>Thank you for your help!</p>

**Comment by securekindness** [comment]
<p>This is a helpful thread, but I’m wondering how to use the id returned. The suggested code returns a list of nodes with base64 enconded ids like this:</p>
<p>{“data”:{“repository”:{“vulnerabilityAlerts”:</p>
<p>{“edges”:[</p>
<p>{“node”:{“id”:"&lt;redacted base64 string here&gt;"}},</p>
<p>{“node”:{“id”:"&lt;redacted base64 string here&gt;"}},</p>
<p>{“node”:{“id”:"&lt;redacted base64 string here&gt;"}}</p>
<p>]}</p>
<p>}}}</p>
<p>How can I use that to get the information from the alert such as the severity, desciption, summary, and other fields listed <a href="https://developer.github.com/v4/object/securityadvisory/#connections" rel="nofollow noopener">https://developer.github.com/v4/object/securityadvisory/#connections</a>? I’m new to graphQL so thanks for bearing with!</p>
